ci/ray_ci/automation/docker_tags_lib.py

```diff
@@ -0,0 +1,253 @@
+import subprocess
+from datetime import datetime
+from typing import List, Optional, Callable
+import os
+import requests
+from dateutil import parser
+
+from ci.ray_ci.utils import logger
+
+bazel_workspace_dir = os.environ.get("BUILD_WORKSPACE_DIRECTORY", "")
+SHA_LENGTH = 6
+
+
+class DockerHubRateLimitException(Exception):
+    """
+    Exception for Docker Hub rate limit exceeded.
+    """
+
+    def __init__(self):
+        super().__init__("429: Rate limit exceeded for Docker Hub.")
+
+
+class RetrieveImageConfigException(Exception):
+    """
+    Exception for failing to retrieve image config.
+    """
+
+    def __init__(self, message: str):
+        super().__init__(f"Failed to retrieve {message}")
+
+
+class AuthTokenException(Exception):
+    """
+    Exception for failing to retrieve auth token.
+    """
+
+    def __init__(self, message: str):
+        super().__init__(f"Failed to retrieve auth token from {message}.")
+
+
+def get_docker_auth_token(namespace: str, repository: str) -> Optional[str]:
```

Comment:
we talked about making these private functions right? should they start with an underscore?

Reply:
Sorry I missed this while fixing the PR last time. Fixed now!